Imix submodule power configuration

shaneleonard opened the pull request:
On imix, there are several pins which need to be 'detached' whenever we switch the power off for a particular submodule. In this case, by 'detached', I mean configured as a GPIO and pulled down, but pin detachment itself is more abstract than that. When the submodule is turned back on, all of these pins need to be restored to their original functions. When testing the power consumption of each submodule, I need to do quite a few of these detachments/restorations in addition to enabling/disabling the power gate for the module, so I wrote a general purpose API that makes this process much easier.

This functionality has other potential uses for the Signpost platform. I put together a rough first pass at an API that could satisfy both my testing needs and the Signpost application, and am making a PR for the opportunity to discuss and make changes.

bradjc commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,117 @@
+trait PowerGated {
+    fn on(&self);
+    fn off(&self);
+}
```
What about just a `power(state: bool)` function?

bradjc commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,117 @@
+    match enabled_modules.rf233 {
+        true => rf233.on(),
+        false => rf233.off(),
+    }
```
Then these just become `rf233.power(enabled_modules.rf233);`.

bradjc commented:
This seems pretty reasonable, and is contained to imix so the API has a limited exposure. Is this enough to merge now, or do you want to wait to either build a more generic version in the kernel, or some sort of userspace control, or test on another platform with power gate-able submodules?

It seems to me that if this is solving your short-term needs then we should merge it.

shaneleonard commented:
I'm fine with merging this for my short-term needs. I think it would be more appropriate to modify this into a generic version if/when there are multiple applications for this type of API.

alevy commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,106 @@
+
+struct ImixSubmodule {
+    gate_pin: &'static GPIOPin,
+    detachable_pins: Option<&'static [DetachablePin]>,
```
Of it's already a list, the list can just be empty instead of wrapping it in an option, which I think makes your implementation a bit more concise.

alevy commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,106 @@
+
+type DetachablePin = (&'static GPIOPin, Option<PeripheralFunction>);
+
+trait Detachable {
```
This trait is not actually being used anywhere, since the only place you use it's methods, you actually know the exact type (`DetachablePin`)

alevy commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+}
+
+trait PowerGated {
```
This trait isn't really used, only it's methods, so may as well just roll them into `ImixSubmodule` and get rid of the trait.

alevy commented on boards/imix/src/power.rs:
```diff
@@ -0,0 +1,106 @@
+    fn power(&self, state: bool);
+}
+
+struct ImixSubmodule {
```
This is a taste thing, but since this type is already namespaced under the `imix` crate, I don't think it's necessary to prefix it's name with `Imix`.

alevy commented on boards/imix/src/main.rs:
```diff
@@ -172,6 +175,14 @@ pub unsafe fn reset_handler() {
+        sensors: true,
+        trng: true,
+    };
+    power::configure_module_power(power_config);
```
Seems like a more natural use of this is to use the `ModulePowerConfig` struct in place of named parameters, so just:

```rust
power::configure_module_power(power::ModulePowerConfig {
    rf233: true,
    nrf51422: true,
    sendosr: true,
    trng: true,
});
```

instead of pulling out the variable.

shaneleonard commented:
My initial reason for breaking out all of traits was to have fine-grained reusability in case there was something specific the Signpost folks wanted to take from this module. However, I think everyone only has at best a vague idea of how this would be relevant to something in Signpost, so it makes the most sense now to just do what's necessary for my purposes.

As for the allocations, it was partly me just compiler-flailing, and partly that I originally thought it could be useful to export the individual submodule structs as global objects. But I think it's probably unnecessary, so I removed it. Hopefully this new commit addresses your comments and clarifies any questions.
